active/Cross Site WebSocket Hijacking.js

```diff
@@ -0,0 +1,100 @@
+ * 
+ * According to RFC 6455, section 10.2 Origin Considerations:
+ * 
+ * "Servers that are not intended to process input from any web page but
+ * only for certain sites SHOULD verify the |Origin| field is an origin
+ * they expect.  If the origin indicated is unacceptable to the server,
+ * then it SHOULD respond to the WebSocket handshake with a reply
+ * containing HTTP 403 Forbidden status code."
+ * 
+ * Which means we can try to repeat WebSocket HTTP Upgrade request with 
+ * a modified Origin header and raise an alert if it was accepted. 
+ * 
+ * Note: Run ajax spider before a scan in order to evaluate your application's 
+ *       JavaScript code which opens WebSocket.
+ * Note: Active scripts are initially disabled, right click the script to enable it.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * 
+ * Author: Piotr Furman <piotr.furman at motorolasolutions.com>
+ */
+
+var Base64 = Java.type("java.util.Base64")
+var Random = Java.type("java.util.Random")
+var String = Java.type("java.lang.String")
+var ByteArray = Java.type("byte[]")
+
+var RISK = 3
+var CONFIDENCE = 2
+var TITLE = "Cross-Site WebSocket Hijacking"
+var DESCRIPTION = "Server accepted WebSocket connection through HTTP Upgrade request with modified Origin header."
+var SOLUTION = "Validate Origin header on WebSocket connection handshake, to ensure only specified origins are allowed to connect. Also, WebSocket handshake should use random tokens, similar to anti CSRF tokens."
+var REFERENCE = "https://tools.ietf.org/html/rfc6455#section-10.2"
+var OTHER = "See also https://portswigger.net/web-security/websockets/cross-site-websocket-hijacking or https://christian-schneider.net/CrossSiteWebSocketHijacking.html"
+var CWEID = 346 // CWE-346: Origin Validation Error, http://cwe.mitre.org/data/definitions/346.html
+var WASCID = 9 // WASC-9 Cross Site Request Forgery, http://projects.webappsec.org/w/page/13246919/Cross%20Site%20Request%20Forgery
+
+function scanNode(as, msg) {
+    var target = msg.getRequestHeader().getURI().toString()
+
+    // check if this is a WebSocket HTTP Upgrade request
+    // request message should include also "Connection: Upgrade" header but let's skip it for now, and switch to msg.isWebSocketUpgrade() if https://github.com/zaproxy/zaproxy/pull/6508 was merged
+    var upgrade_header = msg.getRequestHeader().getHeader("Upgrade")
+    if (!upgrade_header || upgrade_header.toLowerCase() !== "websocket") {
```

Comment:
Should the second part be a contains check instead of not equal? (Just to cover a greater set of circumstances)

Reply:
For WebSocket handshakes "upgrade" header should be equal to "websocket", it won't have any other value. According to specification, handshake should have also a `connection: upgrade` header. And you're right, "connection" might contain multiple entries. 

But in practice, it should be enough if we check just the `Upgrade: websocket` header. AFAIK there's no scenario when the browser would send proper "upgrade" header, and "connection" header without "upgrade" value in the same request. 

And once we switch to `msg.isWebSocketUpgrade()`, proper check will be executed in there.

Reply:
Sounds good, thanks